lib/SPIRV/libSPIRV/SPIRVInstruction.h

```diff
@@ -2721,61 +2721,68 @@ _SPIRV_OP(PowN, true, 10)
 
 class SPIRVAtomicInstBase : public SPIRVInstTemplateBase {
 public:
-  llvm::Optional<ExtensionID> getRequiredExtension() const override {
-    if (getOpCode() == OpAtomicFAddEXT)
-      return ExtensionID::SPV_EXT_shader_atomic_float_add;
-    return {};
-  }
-
   SPIRVCapVec getRequiredCapability() const override {
-    SPIRVCapVec CapVec;
-    if (!hasType())
-      return CapVec;
-
-    // Most atomic instructionsrequire a specific capability when
+    // Most of the atomic instructions require a specific capability when
     // operating on 64-bit integers.
-    if (getType()->isTypeInt(64)) {
-      // In SPIRV 1.2 spec only 2 atomic instructions have no result type:
-      // 1. OpAtomicStore - need to check type of the Value operand
-      // 2. OpAtomicFlagClear - doesn't require Int64Atomics capability.
-      CapVec.push_back(CapabilityInt64Atomics);
-    } else if (getOpCode() == OpAtomicFAddEXT) {
-      if (getType()->isTypeFloat(32))
-        CapVec.push_back(CapabilityAtomicFloat32AddEXT);
-      else if (getType()->isTypeFloat(64))
-        CapVec.push_back(CapabilityAtomicFloat64AddEXT);
-      else
-        llvm_unreachable(
-            "AtomicFAddEXT can only be generated for f32 or f64 types");
-    }
-    // Per the spec OpAtomicCompareExchangeWeak, OpAtomicFlagTestAndSet and
-    // OpAtomicFlagClear instructions require kernel capability. But this
-    // capability should be added by setting OpenCL memory model.
-    return CapVec;
+    // In SPIRV 1.2 spec, only 2 atomic instructions have no result type:
+    // 1. OpAtomicStore - need to check type of the Value operand
+    // 2. OpAtomicFlagClear - doesn't require Int64Atomics capability.
+    // Besides, OpAtomicCompareExchangeWeak, OpAtomicFlagTestAndSet and
+    // OpAtomicFlagClear instructions require the "kernel" capability. But this
+    // capability should be added by setting the OpenCL memory model.
+    if (hasType() && getType()->isTypeInt(64))
+      return {CapabilityInt64Atomics};
+    return {};
   }
 
   // Overriding the following method because of particular OpAtomic*
-  // instructions. We may have to declare additional capabilities,
-  // e.g. based on operand types.
+  // instructions that declare additional capabilities, e.g. based on operand
+  // types.
+  void setOpWords(const std::vector<SPIRVWord> &TheOps) override {
+    SPIRVInstTemplateBase::setOpWords(TheOps);
+    for (auto RC : getRequiredCapability())
+      Module->addCapability(RC);
+  }
+};
+
+class SPIRVAtomicStoreInstBase : public SPIRVAtomicInstBase {
```

Comment:
Why do we have "Base" in `SPIRVAtomicStoreInstBase`, does it have any ancestors?

Reply:
It's a "base template argument" for the instruction type that will be formed as a result of macro unfolding. The names would probably clash was I to rename the class into a simple `SPIRVAtomicStoreInst`.

Reply:
Sorry, what exactly names would clash?

Reply:
Sorry, I didn't realise that `_SPIRV_OP(_SPECIAL)` as is right now would unfold into `SPIRVAtomicStore` instead of `SPIRVAtomicStoreInst`. Please discard the name clash argument.

Personally, I slightly prefer using `Base` due to the second point made in the initial reply (it is a "template base", and `_SPIRV_OP_SPECIAL` expects the `*Base` naming as a second parameter to signify that). However, if this looks overly complicated or misleading for the code reader, I'm ready to change this so that the class is named `SPIRVAtomicStoreInst` instead, and the `*Base` naming requirement is lifted from the macro parameter.

Reply:
Yes, it looks overly complicated and misleading to me.

Reply:
Deal, fixed with the latest revision.